codec/src/main/java/io/netty/handler/codec/MessageToMessageEncoder.java

```diff
@@ -66,68 +64,31 @@ protected MessageToMessageEncoder() {
-     * Returns {@code true} if the given message should be handled. If {@code false} it will be passed to the next
-     * {@link ChannelOutboundHandler} in the {@link ChannelPipeline}.
-     */
-    public boolean acceptOutboundMessage(Object msg) throws Exception {
-        return matcher.match(msg);
+        super(outboundMessageType);
     }
 
     @Override
-    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
-        RecyclableArrayList out = null;
+    protected final void encode(ChannelHandlerContext ctx, I msg, ChannelPromiseAggregatorFactory promise)
+            throws Exception {
+        RecyclableArrayList out = RecyclableArrayList.newInstance();
         try {
-            if (acceptOutboundMessage(msg)) {
-                out = RecyclableArrayList.newInstance();
-                @SuppressWarnings("unchecked")
-                I cast = (I) msg;
-                try {
-                    encode(ctx, cast, out);
-                } finally {
-                    ReferenceCountUtil.release(cast);
-                }
+            encode(ctx, msg, out);
 
-                if (out.isEmpty()) {
-                    out.recycle();
-                    out = null;
+            if (out.isEmpty()) {
+                out.recycle();
+                out = null;
 
-                    throw new EncoderException(
-                            StringUtil.simpleClassName(this) + " must produce at least one message.");
-                }
-            } else {
-                ctx.write(msg, promise);
+                throw new EncoderException(
+                        StringUtil.simpleClassName(this) + " must produce at least one message.");
             }
-        } catch (EncoderException e) {
-            throw e;
-        } catch (Throwable t) {
-            throw new EncoderException(t);
         } finally {
-            if (out != null) {
-                final int sizeMinusOne = out.size() - 1;
-                if (sizeMinusOne == 0) {
-                    ctx.write(out.get(0), promise);
-                } else if (sizeMinusOne > 0) {
-                    // Check if we can use a voidPromise for our extra writes to reduce GC-Pressure
-                    // See https://github.com/netty/netty/issues/2525
-                    ChannelPromise voidPromise = ctx.voidPromise();
-                    boolean isVoidPromise = promise == voidPromise;
-                    for (int i = 0; i < sizeMinusOne; i ++) {
-                        ChannelPromise p;
-                        if (isVoidPromise) {
-                            p = voidPromise;
-                        } else {
-                            p = ctx.newPromise();
-                        }
-                        ctx.write(out.get(i), p);
-                    }
-                    ctx.write(out.get(sizeMinusOne), promise);
-                }
-                out.recycle();
+            if (out == null) {
+                return;
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) Remove this return statement from this finally block. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1143)